server/src/main/java/org/opensearch/index/engine/NRTReplicationReaderManager.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.index.engine;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.lucene.index.DirectoryReader;
+import org.apache.lucene.index.LeafReader;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.index.SegmentInfos;
+import org.apache.lucene.index.SoftDeletesDirectoryReaderWrapper;
+import org.apache.lucene.index.StandardDirectoryReader;
+import org.opensearch.common.lucene.Lucene;
+import org.opensearch.common.lucene.index.OpenSearchDirectoryReader;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * This is an extension of {@link OpenSearchReaderManager} for use with {@link NRTReplicationEngine}.
+ * The manager holds a reference to the latest {@link SegmentInfos} object that is used to refresh a reader.
+ *
+ * @opensearch.internal
+ */
+public class NRTReplicationReaderManager extends OpenSearchReaderManager {
+
+    protected static Logger logger = LogManager.getLogger(NRTReplicationReaderManager.class);
+    private volatile SegmentInfos currentInfos;
+
+    /**
+     * Creates and returns a new SegmentReplicationReaderManager from the given
+     * already-opened {@link OpenSearchDirectoryReader}, stealing
+     * the incoming reference.
+     *
+     * @param reader the SegmentReplicationReaderManager to use for future reopens
+     */
+    NRTReplicationReaderManager(OpenSearchDirectoryReader reader) {
+        super(reader);
+        currentInfos = unwrapStandardReader(reader).getSegmentInfos();
+    }
+
+    @Override
+    protected OpenSearchDirectoryReader refreshIfNeeded(OpenSearchDirectoryReader referenceToRefresh) throws IOException {
+        Objects.requireNonNull(referenceToRefresh);
+        final List<LeafReader> subs = new ArrayList<>();
+        final StandardDirectoryReader standardDirectoryReader = unwrapStandardReader(referenceToRefresh);
+        for (LeafReaderContext ctx : standardDirectoryReader.leaves()) {
+            subs.add(ctx.reader());
+        }
+        DirectoryReader innerReader = StandardDirectoryReader.open(referenceToRefresh.directory(), currentInfos, subs, null);
+        final DirectoryReader softDeletesDirectoryReaderWrapper = new SoftDeletesDirectoryReaderWrapper(
+            innerReader,
+            Lucene.SOFT_DELETES_FIELD
+        );
+        logger.trace("updated to SegmentInfosVersion=" + currentInfos.getVersion() + " reader=" + innerReader);
```

Comment:
Lets use deferred logging via a consumer `() -> doBlah`  for trace level

Reply:
done.